### PR TITLE
【训练营】Checkpoint 读取工具

### PR DESCRIPTION
**Pull request by ArcaLunar**
Checkpoint 读取工具主要参数：

- `--checkpoint_dir` 训练过程中的保存目录
- `--save_steps` 每 N 次保存一次，设置为 0 则不保存
- `--max_checkpoint_keep` 最多保留 K 个 checkpoint
- `--save_optimizer_state` 是否保存优化器的状态
- `--resume_from` 从指定 checkpoint 目录恢复训练

Checkpoint 文件可以通过从 `/data/shared/....../llmc/gpt2` (or `llama3`) 的原始模型参数训练而来，例子可见仓库中的 REPORT.md（Experiment 实际上也测试了llama3，但是命令只记录了 GPT2 训练），`model.bin, optimizer.bin, trainer_state.json` 都可以从训练中获取．因此不在附件中提供

## Experiment

```bash
CUDA_VISIBLE_DEVICES=5,6,7 ./gpt2 --input_bin ../../data/llmc/gpt2/tinyshakespeare/tiny_shakespeare_train.bin --llmc_filepath ../../data/llmc/gpt2/gpt2_124M.bin --checkpoint_dir ../ckpt2/gpt2-noresume/ --num_iteration 100 --save_steps 20 --save_optimizer_state true --max_checkpoint_keep 10
```

```bash
CUDA_VISIBLE_DEVICES=5,6,7 ./gpt2 --input_bin ../../data/llmc/gpt2/tinyshakespeare/tiny_shakespeare_train.bin --llmc_filepath ../../data/llmc/gpt2/gpt2_124M.bin --checkpoint_dir ../ckpt2/gpt2-resumefrom40/ --num_iteration 100 --save_steps 20 --save_optimizer_state true --max_checkpoint_keep 10 --resume_from ../ckpt2/gpt2-noresume/checkpoint_step_000040/ > ../ckpt2/gpt2-resumefrom40/gpt2-resume.log 2>&1
```

（以上两条训练命令同样用 llama3 也运行了）

运行 compare_loss.py，对于 llama3 模型，由于从 step 40 恢复训练，所以 step 1~40 数据缺失，而其余 60 步的 loss 在 FP32, BF16 下均吻合

```text
  Summary: 60/100 steps matched

==================================================
Overall Summary:
  fp32:    0/1 test cases passed (threshold: 1e-05)
  bfloat16: 0/0 test cases passed (threshold: 1e-02)
  Total:   0/1 test cases passed
==================================================

==================================================
Overall Summary:
  fp32:    0/0 test cases passed (threshold: 1e-05)
  bfloat16: 0/1 test cases passed (threshold: 1e-02)
  Total:   0/1 test cases passed
==================================================
```

对于 GPT2，模型保存的逻辑有误：训练中 lm_head 与 wte 并非真共享，而 LLMC 存取又按“共享”假设处理，resume 后 lm_head 很容易和 noresume 不一致。解决方法是把训练用 checkpoint 从 LLMC 回调路径切到原生 StateDict 二进制路径，并在加载后显式重建权重绑定语义 (`example/gpt2/main.cc`)．经过修复后，也可以通过．

### CODE REVIEW

**Comment by ArcaLunar** [comment]
> 1. 请把所有commit信息rebase成一个
> 
> 2. 另外，解决一下当前的代码冲突

已整理 commit 历史（保留上游 merge）+解决冲突

File: example/gpt2/main.cc [hunk 1]
```diff
@@ -320,9 +335,100 @@ void Train(const nn::parallel::Rank &rank) {
 
     auto impl = core::GetDeviceGuardImpl(device.type());
 
+    int start_step = 0;
+    float best_loss = std::numeric_limits<float>::infinity();
+    if (!FLAGS_resume_from.empty()) {
```
Comment (JYMiracle305) [inline example/gpt2/main.cc]: 建议把主流程中恢复、保存、清理旧的Checkpoint提成公共函数，尽量让主流程简洁，另外各个训练入口可以复用。

File: example/llama3/main.cc [hunk 2]
```diff
@@ -73,6 +77,12 @@ DEFINE_uint32(pipeline_parallel, 1, "Pipeline Parallel world size, specified the
 DEFINE_uint32(virtual_pipeline_parallel, 1, "Number of chunks in PP stage.");
 // precision
 DEFINE_string(dtype, "float32", "precision used in training (float32/bfloat16)");
+DEFINE_uint32(save_steps, 0, "save checkpoint every N steps; 0 disables saving");
+DEFINE_string(resume_from, "", "checkpoint directory to resume from");
+DEFINE_string(checkpoint_dir, "./checkpoints", "root directory used to store checkpoints");
+DEFINE_uint32(max_checkpoint_keep, 3, "max number of checkpoint steps to keep");
+DEFINE_bool(save_optimizer_state, true, "whether optimizer state is persisted in checkpoints");
+DEFINE_string(checkpoint_format, "bin", "checkpoint format: bin|pth");
```
[the remark this answers is not in the document]
Comment (ArcaLunar) [reply example/llama3/main.cc]: 其实应该是没有这个 use_llmc_checkpoint_io 的，但是因为 GPT2 `FromLLMC()` 里的 TODO 说明了`wte` 和 `lm_head` 不是真共享权重，而 `FromLLMC()` 不能动因为要读取原始的 `gpt2_124M.bin` 数据．`SaveAsLLMC()` 没有写入 `lm_head` 也是为了和原始 LLMC 文件兼容．然而这样的话，恢复训练就会导致权重不对，从而导致 loss 不一致．所以在 GPT2 训练时加了这个参数用 fallback save & load.

File: example/gpt2/main.cc (hunk 1 above)
Comment (ArcaLunar) [reply example/gpt2/main.cc]: 将 `main.cc` 中的恢复过程提取为 `infini_train::ResumeFromCheckpoint()`，并通过 `std::tie()` 获取 start_step 等信息，通过引用传递实现参数恢复。使用 llama3 进行简单测试，loss 可以复现．

File: example/llama3/main.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (ArcaLunar) [reply example/llama3/main.cc:86]: 已去掉 gpt2/main.cc 中 `use_llmc_checkpoint_io`

File: example/common/utils.cc [hunk 3]
```diff
@@ -61,4 +66,53 @@ void ReadVectorShardFloat(std::ifstream &ifs, float *dst, int64_t len, int64_t s
     ifs.seekg(base + std::streamoff(len * sizeof(float)));
 }
 
+std::tuple<int, float, size_t> ResumeFromCheckpoint(
```
Comment (JYMiracle305) [inline example/common/utils.cc]: 这个参数和返回值都用struct表示吧

File: example/gpt2/main.cc [hunk 4]
```diff
@@ -320,9 +334,69 @@ void Train(const nn::parallel::Rank &rank) {
...
+    CheckpointLoadOptions load_options;
+    load_options.load_optimizer_state = true;
+    load_options.model_bin_loader = [](nn::Module *target_model, const std::filesystem::path &model_path) {
+        auto loaded_model = GPT2::FromLLMC(model_path.string());
+        target_model->LoadStateDict(loaded_model->StateDict());
+    };
+    std::tie(start_step, best_loss, saved_data_batch_idx) = infini_train::ResumeFromCheckpoint(
+        FLAGS_resume_from, rank, model, optimizer, train_loader, state, train_iter, load_options);
+
+    auto save_checkpoint = [&](const std::filesystem::path &save_dir, int64_t global_step,
```
Comment (JYMiracle305) [inline example/gpt2/main.cc]: 这个表达式内部也可以提一个函数（类似 SaveCheckpoint）到utils.cc，内部只构造一个参数的struct，然后调用SaveCheckpoint

Comment (ArcaLunar) [reply example/gpt2/main.cc]: 将 save_checkpoint 的逻辑提取为 `utils.h:SaveCheckpoint()`，主程序中的 lambda 函数只用于获取信息、构造参数并调用函数

File: example/common/utils.cc (hunk 3 above)
Comment (ArcaLunar) [reply example/common/utils.cc]: 已修改，用 struct 打包了输入输出
